tree/treeplayer/src/TSimpleAnalysis.cxx

```diff
@@ -219,41 +239,125 @@ bool TSimpleAnalysis::Run()
+         "empty chain", // -1
+         "invalid entry number", // -2
+         "cannot open the file", // -3
+         "missing tree", // -4
+         "internal error" // -5
+         };
+
+   auto generateHisto = [&](const std::pair<TChain*, TDirectory*>& job){
+      TChain* chain = job.first;
+      TDirectory* taskDir = job.second;
+      taskDir->cd();
+      std::vector<TH1F *> vPtrHisto;
+      for (const auto &histo : fHists) {
+         const std::string& expr = histo.second.first;
+         const std::string& histoName = histo.first;
+         const std::string& cut = histo.second.second;
+
+         chain->Draw((expr + ">>" + histoName).c_str(), cut.c_str(), "goff");
+         TH1F *ptrHisto = (TH1F*)taskDir->Get(histoName.c_str());
+         int errValue;
+         std::string errFile;
+         if (!checkChainLoadResult(chain, errValue, errFile)) {
+            ::Error("TSimpleAnalysis::Run",
+                    "Load failure in file %s: %s", errFile.c_str(), errors[-errValue]);
+            return std::vector<TH1F *>();
+         }
+         vPtrHisto.emplace_back(ptrHisto);
+      }
+      return vPtrHisto;
+   };
+
+#if 1
+
+   ROOT::EnableThreadSafety();
+   ThreadPool pool(8);
+
+   // Do the chain of the fInputFiles
+   std::vector<std::pair<TChain*, TDirectory*>> vChains;
+   TChain *ch;
+   for (size_t i = 0; i < fInputFiles.size(); ++i){
+      const std::string& inputfile = fInputFiles[i];
+      ch = new TChain(fTreeName.c_str());
+      ch->Add(inputfile.c_str());
+      TDirectory* taskDir = gROOT->mkdir(TString::Format("TSimpleAnalysis_taskDir_%d", (int)i));
+      vChains.emplace_back(std::make_pair(ch, taskDir));
+   }
+
+   auto vFileswHists = pool.Map(generateHisto, vChains);
 
-      chain.Draw((expr + ">>" + histoName).c_str(), cut.c_str(), "goff");
-      TH1F *ptrHisto = (TH1F*)gDirectory->Get(histoName.c_str());
-      if (!ptrHisto)
+   // If a file does not exist, one of the vFileswHists
+   // will be a vector of length 0. Detect that.
+   for (auto&& histsOfJob: vFileswHists) {
+      if (histsOfJob.empty())
          return false;
-      ptrHisto->Write();
+   }
+
+   std::vector<TH1F *> vPtrHisto{vFileswHists[0]};
+
+   ofile.cd();
+   for (unsigned j = 0; j < fHists.size(); j++){
+      for (unsigned i = 1; i < vFileswHists.size(); i++){
+         if (!vFileswHists[i][j]) {
+            // ignore that sum histogram:
+            delete vPtrHisto[j];
+            vPtrHisto[j] = nullptr;
+            continue;
+         }
+         if (vPtrHisto[j])
+            vPtrHisto[j]->Add(vFileswHists[i][j]);
+      }
+      if (vPtrHisto[j])
+         vPtrHisto[j]->Write();
    }
    return true;
 }
 
+#else
+
+// Do the chain of the fInputFiles
+TChain* chain = new TChain(fTreeName.c_str());
+for (const std::string& inputfile: fInputFiles)
+   chain->Add(inputfile.c_str());
+
+auto vHisto = generateHisto({chain,gDirectory});
+if (vHisto.empty())
+   return false;
+for (auto histo: vHisto) {
```

Comment:
Before this for (that contains the `Write()`) we need `ofile.cd();` to make sure we write into the output file.